ddtrace/internal/remoteconfig/worker.py

```diff
@@ -93,3 +82,108 @@ def _online(self):
+            atexit.register(self.disable)
+            return True
+        return False
+
+    def start_subscribers(self):
+        # type: () -> None
+        """
+        Disable the remote config service and drop, remote config can be re-enabled
+        by calling ``enable`` again.
+        """
+        self._enable = False
+        log.debug(
+            "[%s][P: %s][P2: %s]  Remote Config Poller fork. Starting Pubsub services",
+            os.getpid(),
+            os.getppid(),
+            self._parent_id,
+        )
+        for pubsub in self._client.get_pubsubs():
+            pubsub.restart_subscriber()
+
+    def _poll_data(self, test_tracer=None):
+        """Force subscribers to poll new data. This function is only used in tests"""
+        for pubsub in self._client.get_pubsubs():
+            pubsub._poll_data(test_tracer=test_tracer)
+
+    def stop_subscribers(self):
+        # type: () -> None
+        """
+        Disable the remote config service and drop, remote config can be re-enabled
+        by calling ``enable`` again.
+        """
+        log.debug(
+            "[%s][P: %s] Remote Config Poller fork. Stopping  Pubsub services",
+            os.getpid(),
+            self._parent_id,
+        )
+        for pubsub in self._client.get_pubsubs():
+            pubsub.stop()
+
+    def disable(self):
+        # type: () -> None
+        self.stop_subscribers()
+        self._client.reset_products()
+
+        if self.status == ServiceStatus.STOPPED:
+            return
+
+        forksafe.unregister(self.start_subscribers)
+        atexit.unregister(self.disable)
+
+        self.stop()
+
+    # def on_shutdown(self):
+    #     self.disable()
+
+    def _stop_service(self, *args, **kwargs):
+        # type: (...) -> None
+        self.stop_subscribers()
+
+        if self.status == ServiceStatus.STOPPED:
+            return
+
+        super(RemoteConfigPoller, self)._stop_service(*args, **kwargs)
+
+    def register(self, product, pubsub_instance, skip_enabled=False):
+        # type: (str, PubSubBase, bool) -> None
+        try:
+            # By enabling on registration we ensure we start the RCM client only
+            # if there is at least one registered product.
+            enabled = True
+            if not skip_enabled:
+                enabled = self.enable()
+            if enabled:
+                self._client.register_product(product, pubsub_instance)
+        except Exception:
+            log.debug("error starting the RCM client", exc_info=True)
+
+    def unregister(self, product):
+        try:
+            self._client.unregister_product(product)
+        except Exception:
+            log.debug("error starting the RCM client", exc_info=True)
+
+    def __enter__(self):
+        # type: () -> RemoteConfigPoller
+        self.enable()
+        return self
+
+    def __exit__(self, *args):
+        # type: (...) -> None
+        self.disable()
+
+
+remoteconfig_poller = RemoteConfigPoller()
```

Comment:
do we have to worry about the case where the main process is only used to spawn another process and then terminates, while the child process keeps doing work? IIUC in this case we wouldn't have a poller running anymore, and a potentially broken pub-sub pipe?

Reply:
Could you give an example? In a Gunicorn application, RC worker is polling in main process data and the child process listening, due to the worker is always running in the main process, if it dies all child workers stop too

Reply:
I'm thinking of [fork off and die](https://wiki.linuxquestions.org/wiki/Fork_off_and_die) for applications that would start a daemon process. Although rare, I wonder if we should ensure support for this too.

Reply:
I checked that with this example:
```
>>> ps -A | grep gunicorn
 452209 pts/3    00:00:00 gunicorn
 452275 pts/3    00:00:00 gunicorn
 452277 pts/3    00:00:00 gunicorn
 452279 pts/3    00:00:00 gunicorn
 452281 pts/3    00:00:00 gunicorn
>>> sudo kill -9 452209 (main process)
>>> ps -A | grep gunicorn
<empty>
```
What's your thoughts? Do you think it's enough?

Reply:
If I understand correctly, I'm testing that behavior on this test:
https://github.com/DataDog/dd-trace-py/pull/5464/files#diff-213a7c2bf1bb1f526312e4a698d5f7118cf7fe692dd4ebaa3ff7de5209552e79R49

A long time ago (in a commit far, far away...) I had an error on `_stop_service` and that test never finished. It was related to your concern: the main process dies but the child process still running

Do you think this test is enough to validate that? do you think we need a specific test?

Reply:
If I understand the test correctly, this is still an example of a server that spawns worker processes and remains in the background. What I would like to validate is the scenario where a process is used to spawn a daemon process which keeps running in the background, while the parent terminates. My understanding is that, with the current design, the poller that only runs in the parent process will die, and the child daemon process will be left with no poller to listen to.

Reply:
Ok, I think I understand now your concerns :smile: 

What do you think about this approach?

Actual behavior of this branch:
```
- Main process -> RC publisher
   - Child 1 -> Subscriber 1
   - ...
   - Child N -> Subscriber N
```

We could create a env var such as `DD_REMOTE_CONFIG_FORK_AND_DIE_SAFE` and if it's True, application runs Remote Config as it is doing now in `1.x`, that means, application runs a Remote Config Worker on each process:
```
- Main process -> RC publisher 1 Subscriber 1
   - Child 1 -> RC publisher 2 Subscriber 2
   - ...
   - Child N -> RC publisher N Subscriber N
```

Reply:
Yep this approach would definitely work. How hard would it be to implement a mechanism that would allow the remaining children to detect that the publisher if offline and negotiate who becomes the new publisher?